src/adapter/src/coord/sequencer/alter_set_cluster.rs

```diff
@@ -7,29 +7,188 @@
+        let (mut stmt, resolved_ids) = mz_sql::names::resolve(&session_catalog, stmt)?;
+
+        match &mut stmt {
+            Statement::CreateMaterializedView(s) => {
+                // Planning for materialized views fails if the object already exists.
+                // Force to skip behavior, which is OK because it still lowers the query.
+                // We need to be careful to not save the new behavior in the catalog. Planning
+                // changes this to Error.
+                s.if_exists = IfExistsBehavior::Skip;
+            }
+            _ => {}
+        }
+
+        let plan = mz_sql::plan::plan(
+            None,
+            &session_catalog,
+            stmt,
+            &Params::empty(),
+            &resolved_ids,
+        )?;
+
+        match entry.item() {
+            CatalogItem::Index(_old_index) => Err(AdapterError::Unsupported("ALTER SET CLUSTER")),
+            CatalogItem::MaterializedView(old_mv) => {
+                if !self.is_compute_cluster(set_cluster) {
+                    let cluster_name = self.catalog().get_cluster(set_cluster).name.clone();
+                    return Err(AdapterError::BadItemInStorageCluster { cluster_name });
+                }
+
+                let old_cluster_id = old_mv.cluster_id;
+
+                let (optimized_expr, desc, create_sql) = match plan {
+                    Plan::CreateMaterializedView(CreateMaterializedViewPlan {
+                        materialized_view,
+                        ..
+                    }) => {
+                        let optimizer =
+                            Optimizer::logical_optimizer(&mz_transform::typecheck::empty_context());
+                        let optimized_expr = optimizer.optimize(materialized_view.expr)?;
+                        let desc =
+                            RelationDesc::new(optimized_expr.typ(), materialized_view.column_names);
+                        assert_eq!(materialized_view.cluster_id, set_cluster);
+
+                        (optimized_expr, desc, materialized_view.create_sql)
+                    }
+                    _ => {
+                        return Err(catalog::Error::new(catalog::ErrorKind::Corruption {
+                            detail: "catalog entry generated inappropriate plan".to_string(),
+                        })
+                        .into())
+                    }
+                };
+
+                let item = CatalogItem::MaterializedView(catalog::MaterializedView {
+                    create_sql,
+                    optimized_expr,
+                    desc,
+                    resolved_ids,
+                    cluster_id: set_cluster,
+                });
+
+                // Allocate a unique ID that can be used by the dataflow builder to
+                // connect the view dataflow to the storage sink.
+                let internal_view_id = self.allocate_transient_id()?;
+                let op = catalog::Op::AlterSetCluster { id, item };
```

Comment:
Should we use `Op::UpdateItem` instead? It would cause a different entry in the audit log that doesn't highlight how we move an item between clusters. @mjibson might have a take!

Reply:
I think this op is fine.